STM32F7 SoC family and STM32F746G Discovery board basic support

pic16f887 opened the pull request:
@erwango, @ydamigos, @dbkinder

I have closed pull request #7220 and created a new one. I had to drop previous PR because it was a mess which is hard to resolve to me (I am newbie in git and GitHub). Sorry for inconvenience.

pic16f887 commented on boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst:
```diff
@@ -0,0 +1,199 @@
+
+Other hardware features are not yet supported on Zephyr porting.
+
+The default configuration can be found in the defconfig file:
```
@dbkinder, I have fixed it.

dbkinder commented on boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst:
```diff
@@ -0,0 +1,199 @@
+
+More information about STM32F746NGH6 can be found here:
+
+       - `STM32F746NGH6 on www.st.com`_
```
These two lines still need to be not intended...

pic16f887 replied:
Shippable prints warnings like this:
"B4 Second line is not empty". Does this mean that commit message body should have only one line?
Also there is one failed test, but I don't understand how check which exactly test was failed. Are these tests from sanity check?

dbkinder reviewed (changes requested):
One more edit, and you've got some problems reported by shippable:  https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/14153/5/tests

erwango commented on arch/arm/soc/st_stm32/stm32f7/soc.h:
```diff
@@ -0,0 +1,50 @@
+#endif /* CONFIG_CLOCK_CONTROL_STM32_CUBE */
+
+#ifdef CONFIG_SERIAL_HAS_DRIVER
+#include <stm32f7xx_ll_usart.h>
```
Move uart.h inclusion to the commit where you introduce uart driver

erwango commented on arch/arm/soc/st_stm32/stm32f7/soc.h:
```diff
@@ -0,0 +1,50 @@
+
+#include "soc_irq.h"
+
+#ifdef CONFIG_CLOCK_CONTROL_STM32_CUBE
```
Move this section to the commit where you introduce clock control driver

erwango reviewed (changes requested):
Additional remarks but it goes in good direction.

General points:
- Add at least a one line comment in your commit messages, otherwise CI will fail
- Enhance commit separation (lot of gpio, clock, uart related code ) in initial commit

Specific for Clock control:
You need to take car of the code under ifdef F4X in stm32_ll_clock.c, it is applicable to F7 as well.

erwango commented:
Seems we've lost some of your commits ..

pic16f887 commented:
I dropped old commits. I attached only base  STM32F7 commit. I will attach rest of updated commits soon.

Commit:
drivers: pinmux: stm32: STM32F7 pinmux support

This patch adds pinmux header file for STM32F7 famlily
microcontrollers

Signed-off-by: Yurii Hamann <yurii@hamann.site>

pic16f887 commented:
I attached commented updated code.

tagunil commented:
@erwango, what is the policy for IRQ handlers for EXTI >= 16? F4 (and now F7) port enables them in stm32_exti_enable, other families do not. Which way should it be unified?

dbkinder commented on boards/arm/stm32f746g_disco/doc/stm32f746g_disco.rst:
```diff
@@ -0,0 +1,199 @@
+- On-board ST-LINK/V2-1 supporting USB re-enumeration capability
+- Five power supply options:
+
+    - ST LINK/V2-1
```
The sublist should be indented to the first character of the text of the parent list:

    - Five power supply options:

      - ST LINK/V2-1

If you indent it too much, it changes the way the sublist is rendered.

pic16f887 replied:
I have fixed it and updated the commit.

dbkinder reviewed (changes requested):
one more edit and we're good to go

dbkinder reviewed (approved):
Thanks!

pic16f887 commented:
@erwango,
Could you please review my latest changes?

ydamigos commented on arch/arm/soc/st_stm32/stm32f7/soc.c:
```diff
@@ -0,0 +1,62 @@
+ *        Tick interrupt priority is not used
+ * @return HAL status
+ */
+uint32_t HAL_GetTick(void)
```
Could you remove this function? It is now provided by arch/arm/soc/st_stm32/common/stm32cube_hal.c

erwango commented on boards/arm/stm32f746g_disco/Kconfig.defconfig:
```diff
@@ -0,0 +1,28 @@
+
+endif # UART_CONSOLE
+
+if GPIO_STM32
```
Move this to SoC (GPIO Commit)

pic16f887 replied:
Why do we need to mix board and SoC code in commit?

erwango replied:
Good question.
In order to avoid confusing variety of GPIO configurations in various boards, it has been decided to enable all GPIO ports as default board configuration (documented here: http://docs.zephyrproject.org/porting/board_porting.html#default-board-configuration)
Then, since all available GPIO are enabled, this no longer depends on board, and can be moved to SoC.
Then, if user wants to deactivate some port as part of use case optimization (footprint and power), this should be done as application configuration.

pic16f887 replied:
How should new board be added?
It's necessary to create one commit with board basic support and another commit that enables GPIO?

erwango replied:
Let's activate GPIOs as part of the GPIO commit.
GPIOs A, B and C are already enabled in arch/arm/soc/st_stm32/common/Kconfig.defconfig.series. Do nothing for these.
In arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.series, enable ports D E F G H and I as they seems to be always available in all F7 SoCs
Then, in matching Kconfig.defconfig.stm32f7xxxx file, enable the additional GPIOs if any (J K)

pic16f887 replied:
Now I have understood what you meant. Thank you.

erwango reviewed (changes requested):
Last comments, we're getting closer!

erwango commented on boards/arm/stm32f746g_disco/pinmux.c:
```diff
@@ -0,0 +1,37 @@
+#endif	/* CONFIG_UART_STM32_PORT_1 */
+#ifdef CONFIG_UART_STM32_PORT_6
+	{STM32_PIN_PC6, STM32F7_PINMUX_FUNC_PC6_USART6_TX},
+	{STM32_PIN_PG7, STM32F7_PINMUX_FUNC_PC7_USART6_RX},
```
PG7 > PC7?

pic16f887 replied:
Thank you. I will check it.

erwango commented on boards/arm/stm32f746g_disco/stm32f746g_disco.dts:
```diff
@@ -0,0 +1,52 @@
+	status = "ok";
+};
+
+&usart6 {
```
Out of curiosity, why do you choose to configure this USART6 in particular?

Question is because we're trying to enforce default board configuration rules (cf http://docs.zephyrproject.org/porting/board_porting.html#default-board-configuration, and #7151). So what would be good is to configure adruino uart (if possible), and if this is for other purpose, just let me know.

pic16f887 replied:
I have chosen USART6 because it's mapped to Arduino connector on this board.

erwango reviewed (changes requested):
Last comment about board configuration, otherwise sounds good

erwango commented on dts/arm/st/stm32f7.dtsi:
```diff
@@ -0,0 +1,202 @@
+		rcc: rcc@40023800 {
+			compatible = "st,stm32-rcc";
+			clocks-controller;
+			#clocks-cells = <2>;
```
Rename to "#clock-cells" (This has changed recently, there is now a warning associated).

erwango commented on arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,31 @@
+
+if SOC_SERIES_STM32F7X
+
+source "arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.stm32f7*"
```
Replace source with gsource, this is causing CI to fail.
This is due to new commit on master

pic16f887 replied:
I have fixed it and updated the commits.

erwango reviewed (changes requested):
2 comments due to recent introduction in master branch

erwango commented:
@pic16f887 , still one comment to be addressed (clocks-cells -> clock-cells), otherwise should be ok.

erwango commented:
@pic16f887 : Last point , can you indicate the tests you've ran?

pic16f887 commented:
@erwango,
It's my mistake. I applied wrong patch with device tree. I will fix it. 
I tested board on "hello world" and "blinky" examples.

pic16f887 commented:
@erwango 
I have updated commit with device tree

erwango commented on boards/arm/stm32f746g_disco/stm32f746g_disco_defconfig:
```diff
@@ -0,0 +1,39 @@
+CONFIG_CLOCK_STM32_SYSCLK_SRC_PLL=y
+# use HSE as PLL input
+CONFIG_CLOCK_STM32_PLL_SRC_HSE=y
+# produce 180MHz clock at PLL output
```
Replace 180 by 260 and we're good to go

pic16f887 replied:
I am not sure about 260 MHz. The  maximum frequency for STM32F746G is 216 MHz.

erwango replied:
you're right, 216 will do

pic16f887 commented:
@erwango,
I will push it a little bit later. I have trouble with my forked repository.

erwango commented:
> I will push it a little bit later. I have trouble with my forked repository.

Ok, sure. Thanks

erwango commented on arch/arm/soc/st_stm32/stm32f7/Kconfig.series:
```diff
@@ -0,0 +1,19 @@
+	select SOC_FAMILY_STM32
+	select SYS_POWER_LOW_POWER_STATE_SUPPORTED
+	select HAS_STM32CUBE
+	select CPU_HAS_MPU
```
Sorry for late catch, but I'm not sure about activating MPU right now.
MPU for Cortex M7 has specific options which might affect performance.
Hence I prefer this is activated specific check/study rather than by default.

pic16f887 replied:
OK. I will deactivate it.

pic16f887 replied:
I have deactivated MPU and updated the commit.

aurel32 commented on arch/arm/soc/st_stm32/stm32f7/dts.fixup:
```diff
@@ -1,5 +1,53 @@
+#define CONFIG_UART_STM32_PORT_1_IRQ_PRI	ST_STM32_USART_40011000_IRQ_0_PRIORITY
+#define CONFIG_UART_STM32_PORT_1_NAME		ST_STM32_USART_40011000_LABEL
+#define PORT_1_IRQ				ST_STM32_USART_40011000_IRQ_0
+
```
I just gave it a try, and it doesn't seem to compile. It seems it should be renamed from `CONFIG_UART_STM32_PORT_1_` to `CONFIG_UART_STM32_USART_1_`

pic16f887 replied:
I am using UART1 on STM32746G Discovery board. It compiles and works fine.

aurel32 replied:
As I said above, I tried your patchset rebased on master, that's why I have the issue and you don't. I have identified that it has been introduced by the following patch which has been recently committed: https://github.com/zephyrproject-rtos/zephyr/commit/fd26514a4d40bc54be40bb294bd03b105d674c40

You will need to fix this issue anyway before the patchset can be merged. The changes are as simple as replacing `_PORT_` by `_USART_` in `dts.fixup`.

pic16f887 replied:
I will check it.

aurel32 replied:
I have rebased your patchset onto master and fixed this issue in the stm32f7 branch of my repository. The patch is available there: https://github.com/aurel32/zephyr/commit/1857f6628d6cdee3930d7aeb66d30b053f8d5e78

Please feel free to just take it and squash it into your commit adding `dts.fixup`.

aurel32 commented on dts/arm/st/stm32f7.dtsi:
```diff
@@ -0,0 +1,202 @@
+	reg = <0x08000000 DT_FLASH_SIZE>;
+	write-block-size = <1>;
+	};
+
```
It's really nitpicking, but the indentation seems wrong here.

aurel32 commented on dts/arm/st/stm32f7.dtsi:
```diff
@@ -0,0 +1,202 @@
+				clocks = <&rcc STM32_CLOCK_BUS_AHB1 0x00000001>;
+				label = "GPIOA";
+		};
+
```
The indentation is also wrong here, compared to the rcc one.

aurel32 replied:
Is there a reason why the all the gpio entries are not included in pinctrl entry like on other STM32 SoCs? I guess that it also explains the indentation difference here.

pic16f887 replied:
I am planning to add pinctrl in next PR. STM32 boards still use pinconf array instead of device tree definitions.

erwango replied:
AFAIK, no boards today uses device tree definitions. There is some work on going, but will take some time (pinctrl API introduction then device tree based generation).
Meanwhile, indeed, you'd better move gpio nodes under pinctrl node, like other STM32 SoCs

pic16f887 replied:
OK.

aurel32 commented:
Thanks for this patch set. I have pointed a few small things, note however some of them might be due to the fact I rebased your patchset over master (that you'll need to do that anyway to fix the `dts/arm/st/mem.h` conflict.

I have been able to use your patch set to add basic support for the STM32F723E-DISCO board, and I have been able to get Zephyr booting on it. Big thanks.

aurel32 reviewed (changes requested):
Thanks for this patch set. I have pointed a few small things, note however some of them might be due to the fact I rebased your patchset over master (that you'll need to do that anyway to fix the dts/arm/st/mem.h conflict.

I have been able to use your patch set to add basic support for the STM32F723E-DISCO board, and I have been able to get Zephyr booting on it. Big thanks.

aurel32 commented on arch/arm/soc/st_stm32/stm32f7/soc.h:
```diff
@@ -0,0 +1,30 @@
+
+#endif /* !_ASMLANGUAGE */
+
+#endif /* _STM3274_SOC_H_ */
```
Given you need to respin the patchset, a small typo here: _STM3274_SOC_H_ => _STM32F7_SOC_H_

pic16f887 replied:
Thank you. I will fix it.

alexanderwachter commented:
Hi, can't you just copy the drivers/pinmux/stm32/pinmux_stm32f4.h file and rename everything from STM32F4_PINMUX_FUNC_* to STM32F7_PINMUX_FUNC_* ?

aurel32 commented:
I have used this patchset to add support for the STM32F723E-DISCO board, the result is published in the stm32f7 branch of my repository: https://github.com/aurel32/zephyr/commits/stm32f7 . I'll do a proper pull request once the current one is merged.

It also adds support for TRNG, USB OTG and Cortex-M7 caches.

pic16f887 commented:
@alexanderwachter

I did it initially but I was requested to left only UART. Basic board support should have only GPIO and UART. Other peripherals should added in additional commits.

pic16f887 commented:
@aurel32
I initially added TRNG, USB OTG, SPI and other peripherals but was requested to delete it for patch with basic support. Now I have a holidays so I will able to fix and rebase my patch on Monday. Then you will be able to add patches with another peripherals.

alexanderwachter commented:
@pic16f887 will you add this in an additional PR?
I'm currently using your parch to add support for the STM32f746ZG nucleo board.
Thanks for your work!

pic16f887 commented:
@alexanderwachter
Yes, I will. I am going to add SPI, I2C and some other peripherals in additional PR.

aurel32 commented on arch/arm/soc/st_stm32/stm32f7/flash_registers.h:
```diff
@@ -0,0 +1,45 @@
+		u32_t latency :4 __packed;
+		u32_t rsvd__4_7 :4 __packed;
+		u32_t prften :1 __packed;
+		u32_t icen :1 __packed;
```
This seems to have been copied from the STM32F4 structure. On the STM32F7, the caches are included in the Cortex-M7, so the fields for controlling the caches are not there. This should be instead:

```
union __flash_acr {
        u32_t val;
        struct {
                u32_t latency :4 __packed;
                u32_t rsvd__4_7 :4 __packed;
                u32_t prften :1 __packed;
                u32_t arten :1 __packed;
                u32_t rsvd__10 :1 __packed;
                u32_t artrst :1 __packed;
                u32_t rsvd__12_31 :20 __packed;
        } bit;
};

```

pic16f887 replied:
I will check it.

pic16f887 replied:
I will check it.

pic16f887 replied:
Thank you. You are right.

Commit:
drivers: flash: stm32: STM32F7 flash memory suport

Added flash memory support for STM32F7 family microcontrollers

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
boards: arm: stm32: Basic support for STM32F746G-DISCO board

The patch contains basic code support, documentation, dts
file and openocd configuration for STM32F746G discovery board.
This patch related to the issue #6982.

Signed-off-by: Yurii Hamann <yurii@hamann.site>

pic16f887 commented:
@erwango 
Could you review my latest changes?

erwango commented:
@pic16f887 , seems GPIO dts update has not been included in last push.

FYI, I'm trying to push some changes that might impact you (https://github.com/zephyrproject-rtos/zephyr/pull/7997). If merged before yours, you'll have to rebase and include them. And in other case, I'll have to do it :-)

pic16f887 commented:
@erwango,
Ok. 
I will be able to update device in the evening.

aurel32 commented on arch/arm/soc/st_stm32/stm32f7/dts.fixup:
```diff
@@ -1,5 +1,53 @@
+
+#define CONFIG_UART_STM32_USART_8_BASE_ADDRESS	ST_STM32_USART_40007C00_BASE_ADDRESS
+#define CONFIG_UART_STM32_PORT_8_BAUD_RATE	ST_STM32_USART_40007C00_CURRENT_SPEED
+#define CONFIG_UART_STM32_USART_8_IRQ_PRI	ST_STM32_USART_40007C00_IRQ_0_PRIORITY
```
This should be CONFIG_UART_STM32_USART_8_BAUD_RATE instead of _PORT_.

pic16f887 replied:
Thank you. I have fixed it.

aurel32 commented:
Thanks for this new version. Beside the small issue regarding USART port 8, it looks all fine to me.

aurel32 reviewed (approved):
Thanks, it now looks all good!

erwango commented on arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.stm32f746xg:
```diff
@@ -0,0 +1,18 @@
+
+config SOC
+	string
+	default stm32f746xx
```
You should use double quotes here: "stm32f476xx".
Seems a recent change from master

erwango commented on arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.series:
```diff
@@ -0,0 +1,15 @@
+gsource "arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.stm32f7*"
+
+config SOC_SERIES
+	default stm32f7
```
Use double quotes: "stm32f7"

erwango commented on dts/arm/st/stm32f7.dtsi:
```diff
@@ -0,0 +1,210 @@
+			compatible = "st,stm32-pinmux";
+			#address-cells = <1>;
+			#size-cells = <1>;
+			reg = <0x48000000 0x2400>;
```
reg = <0x40020000 0x8000>;

pic16f887 replied:
Why size is 0x8000 for pin-controller? As I understand it's size of all GPIO sections. 8 GPIO sections has size 0x2400.

erwango replied:
My bad, I mixed pinmux and uart address.
2400 is fine. Removing my comment. Thanks

erwango commented on dts/arm/st/stm32f7.dtsi:
```diff
@@ -0,0 +1,210 @@
+			label = "UART_6";
+		};
+
+		usart7: serial@40007800 {
```
Update to UART instead of USART

erwango commented on dts/arm/st/stm32f7.dtsi:
```diff
@@ -0,0 +1,210 @@
+		};
+
+		usart7: serial@40007800 {
+			compatible = "st,stm32-usart", "st,stm32-uart";
```
Remove USART compatible

erwango commented on dts/arm/st/stm32f7.dtsi:
```diff
@@ -0,0 +1,210 @@
+		usart7: serial@40007800 {
+			compatible = "st,stm32-usart", "st,stm32-uart";
+			reg = <0x40007800 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00000020>;
```
clocks = <&rcc STM32_CLOCK_BUS_APB1 0x40000000>

erwango commented on dts/arm/st/stm32f7.dtsi:
```diff
@@ -0,0 +1,210 @@
+		usart8: serial@40007C00 {
+			compatible = "st,stm32-usart", "st,stm32-uart";
+			reg = <0x40007C00 0x400>;
+			clocks = <&rcc STM32_CLOCK_BUS_APB2 0x00000020>;
```
clocks = <&rcc STM32_CLOCK_BUS_APB1 0x80000000>

erwango commented on dts/arm/st/stm32f7.dtsi:
```diff
@@ -0,0 +1,210 @@
+			label = "UART_7";
+		};
+
+		usart8: serial@40007C00 {
```
UART

erwango commented on dts/arm/st/stm32f7.dtsi:
```diff
@@ -0,0 +1,210 @@
+		};
+
+		usart8: serial@40007C00 {
+			compatible = "st,stm32-usart", "st,stm32-uart";
```
Remove USART compatible

erwango reviewed (changes requested):
Some additional comments, but we're getting close.
Thanks for your patience (anyway, we'll have to wait until V1.12 is released before being able to merge)

Commit:
arm: stm32: STM32F7 family device tree

This patch includes:
STM32F7 family device tree file with basic and UART definitions.
STM32F746 subfamily device tree file.
Memory definitions for STM32F746xG subfamily.

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
arch: arm: stm32: Basic STM32F7 family support

The patch includes support for STM32F746xG subfamily.
Related to issue #6981.

Signed-off-by: Yurii Hamann <yurii@hamann.site>

Commit:
drivers: clock_control: STM32F7 family clock control

This patch adds clock control support for STM32F7 family
microcontrollers.

Signed-off-by: Yurii Hamann <yurii@hamann.site>

pic16f887 commented:
@erwango 
I have rebased the code.

aurel32 commented on drivers/interrupt_controller/exti_stm32.c:
```diff
@@ -521,6 +559,59 @@ static void __stm32_exti_connect_irqs(struct device *dev)
 		CONFIG_EXTI_STM32_RTC_WKUP_IRQ_PRI,
 		__stm32_exti_isr_22, DEVICE_GET(exti_stm32),
 		0);
++#elif CONFIG_SOC_SERIES_STM32F7X
```
The shippable build failed due to this leftover from the rebase. Same issue a few lines below.

pic16f887 replied:
Thank you. I have update the code.

Commit:
drivers: interrupt_controller: stm32: STM32F7 EXTI support

This patch adds EXTI support for STM32F7 family
microcontrollers.

Signed-off-by: Yurii Hamann <yurii@hamann.site>

aurel32 commented on arch/arm/soc/st_stm32/stm32f7/Kconfig.defconfig.stm32f746xg:
```diff
@@ -0,0 +1,28 @@
+config GPIO_STM32_PORTK
+	default y
+
++endif # GPIO_STM32
```
I just realized there is still a small rebase issue, there is a `+` before the `endif`.

pic16f887 replied:
Thank you. I have updated the code.

Commit:
drivers: gpio: stm32: STM32F7 GPIO support

This patch adds GPIO support for STM32F7 family microcontrollers.

Signed-off-by: Yurii Hamann <yurii@hamann.site>

aurel32 commented:
Hi @pic16f887, following the merge of #8359, this patchset needs yet another rebase :-(. In the patch `serial: stm32: STM32F7 UART support`, the part changing `drivers/serial/uart_stm32.c` should simply be dropped.

Commit:
serial: stm32: STM32F7 UART support

The patch adds serial driver support for STM32F7 family
microcontrollers, includes pinmux definitions and DTS fixup file.

Signed-off-by: Yurii Hamann <yurii@hamann.site>

pic16f887 commented:
Hi @aurel32,
I have rebased the code.

aurel32 commented:
Hi @pic16f887. Thanks for the rebase, I have just checked, it's all good.
